#!/usr/bin/env python3

import os

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy

from nav_msgs.msg import Odometry
from geometry_msgs.msg import Pose, Quaternion, Vector3, Point
from visualization_msgs.msg import Marker, MarkerArray
from std_msgs.msg import Header, ColorRGBA

from transforms3d.euler import quat2euler

import csv
import numpy as np

from ament_index_python.packages import get_package_share_directory

from scipy.interpolate import CubicSpline

from sbamp.utils import load_waypoints

class VisualizeNode(Node):
    def __init__(self):
        super().__init__("visualize_node")
        self.get_logger().info("Visualize Node Launched")

        self.declare_parameter('waypoint_file_name', 'waypoints_levine.csv')
        self.declare_parameter('visualize_wp_topic', '/visualization/waypoints')

        waypoint_file_name = self.get_parameter('waypoint_file_name').get_parameter_value().string_value
        visualize_wp_topic = self.get_parameter('visualize_wp_topic').get_parameter_value().string_value

        package_share_dir = get_package_share_directory("sbamp")

        waypoint_file_path = os.path.join(package_share_dir, 'config', waypoint_file_name)
        self.waypoints = np.array(load_waypoints(waypoint_file_path))

        qos_profile = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE,
            depth=10
        )   

        # Subscribers

        # Publishers
        self.waypoint_marker_publisher_ = self.create_publisher(MarkerArray, visualize_wp_topic, qos_profile)
        
        # self.visualization_timer = self.create_timer(1, self.visualize_waypoints)

        # Visualize the waypoints
        self.visualize_waypoints()

        self.prev_valid_index = 0
        self.prev_ego_waypoints = []
    
    def visualize_waypoints(self):
        marker_array = MarkerArray()
        for i, wp in enumerate(self.waypoints):
            x, y, yaw, qw, qx, qy, qz = wp

            marker = Marker()
            marker.header = Header()
            marker.header.frame_id = "map"
            marker.header.stamp = self.get_clock().now().to_msg()
            marker.id = i
            marker.type = Marker.SPHERE
            marker.action = Marker.ADD
            
            marker.pose = Pose()
            marker.pose.position = Point(x=x, y=y, z=0.0)
            marker.pose.orientation = Quaternion(x=qx, y=qy, z=qz, w=qw)

            marker.scale = Vector3(x=0.1, y=0.1, z=0.1)
            marker.color = ColorRGBA(r=1.0, g=1.0, b=0.0, a=0.9)

            marker.lifetime.sec = 0

            marker_array.markers.append(marker)

        self.waypoint_marker_publisher_.publish(marker_array)
        self.get_logger().info("Waypoints Visualized")

def main(args=None):
    rclpy.init(args=args)
    node = VisualizeNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()